solr/core/src/java/org/apache/solr/core/TransientSolrCoreCacheDefault.java

```diff
@@ -89,6 +86,21 @@ public TransientSolrCoreCacheDefault(CoreContainer coreContainer) {
     transientDescriptors = new LinkedHashMap<>(initialCapacity);
   }
 
+  private void onEvict(SolrCore core) {
+    if (core.getOpenCount() > 1) {
```

Comment:
@ben-manes is there a way to check this condition (core.getOpenCount() <= 1) during the selection of the entry to evict in Caffeine cache?
Here we put back the evicted entry if this condition is not satisfied, potentially going beyond the max size defined, whereas there may be other potential entries to evict for which this condition is satisfied.

Reply:
The way to handle this is to [pin entries](https://github.com/ben-manes/caffeine/wiki/Faq#pinning-entries). When an entry should be pinned you can update it through the cache (e.g. `cache.asMap.compute`) and have a `Weigher` to evaluate the entry's size as `0`. Then it won't take any capacity and be skipped over during size eviction. Similarly for expiration using an `Expiry` that gives an infinite lifetime.

This approach is preferred because then the cache won't surprisingly violate the contracts, won't perform O(n) scans searching for candidates, and the decisions are more explicit. Caffeine skips over zero-weight entries, but in the future it could move them elsewhere. So far this pin/unpin approach seems to clearer than Ehcache's [EvictionAdvisor](https://github.com/ehcache/ehcache3/blob/master/ehcache-api/src/main/java/org/ehcache/config/EvictionAdvisor.java) (a soft no; will evict an entry [regardless](https://github.com/ehcache/ehcache3/blob/5cfa96b868423f0fbb19bf27ead647823b14a165/ehcache-impl/src/main/java/org/ehcache/impl/internal/store/heap/OnHeapStore.java#L1578-L1583) of the advice) and Coherence's [EvictionApprover](https://github.com/oracle/coherence/blob/d97c16ca8069143a8117356914da1bc341e8c19f/prj/coherence-core/src/main/java/com/tangosol/net/cache/ConfigurableCacheMap.java#L331-L364) is a hard no (which could lead to heap exhaustion), but please let me know if you run into issues that advocate for reevaluating our approach.

Reply:
From the pinning docs:

> The weight and expiration are evaluated when the entry is written into the cache.

This suggests that if the weight changes, then it needs to be re-put into the cache?  For SolrCores in this cache, I think it would be awkward/entangled to do this.  Basically whenever anyone requests or returns a core, we call this method.  Not much lines-of-code but it's conceptually strange and readers of the code and our future selves will wonder what's going on.

Any way, I think the approach here (put back on eviction) is expected to be rare.  Basically, the least requested item (probably hasn't been requested in a while) is still being used, which suggests some long-running operation on it is underway.  Also I have some logs on this so we can observe if this is happening more often.

Reply:
That's make sense. I think informing the cache that the entry's state changed is desirable for managing resources, but understandably hard as projects evolve without that behavior in mind. If you obtain metrics that indicate a problem then it will be interesting to discuss our options. I think your approach is pretty good as a workaround and sounds safe enough.